src/Control/Exception/Annotated.hs

```diff
@@ -263,83 +280,115 @@ tryFlatten exn =
+
+-- | Adds only the current 'CallStack' to the checkpoint. This function searches
+-- any thrown exception for a pre-existing 'CallStack' and will merge the given
+-- pre-existing 'CallStack' with the one on this function, in an attempt to
+-- preserve the actual call history.
 --
 -- @since 0.1.0.0
 checkpointCallStack
     :: (MonadCatch m, HasCallStack)
     => m a
     -> m a
 checkpointCallStack =
-    checkpointCallStackWith []
+    withFrozenCallStack (checkpoint (Annotation callStack))
 
 -- | Add the list of 'Annotation' to any exception thrown in the following
 -- action.
 --
 -- @since 0.1.0.0
-checkpointMany :: (MonadCatch m) => [Annotation] -> m a -> m a
-checkpointMany ann action =
+checkpointMany :: (MonadCatch m, HasCallStack) => [Annotation] -> m a -> m a
+checkpointMany anns action =
     action `Safe.catch` \(exn :: SomeException) ->
-        Safe.throw . annotate ann $ case Safe.fromException exn of
-            Just (e' :: AnnotatedException SomeException) ->
-                e'
-            Nothing -> do
-                new exn
+        Safe.throw
+            . addCallStackToException callStack
+            . annotate anns
+            $ case Safe.fromException exn of
+                Just (e' :: AnnotatedException SomeException) ->
+                    e'
+                Nothing -> do
+                    pure exn
 
 -- | Retrieves the 'CallStack' from an 'AnnotatedException' if one is present.
 --
+-- The library maintains an internal check that a single 'CallStack' is present
+-- in the list, so this only returns the first one found. If you have added
+-- a 'CallStack' directly to the @['Annotation']@, then this will likely break.
+--
 -- @since 0.1.0.0
 annotatedExceptionCallStack :: AnnotatedException exception -> Maybe CallStack
 annotatedExceptionCallStack exn =
-    let (stacks, _rest) = callStackInAnnotations (annotations exn)
+    let (stacks, _rest) = tryAnnotations (annotations exn)
     in listToMaybe stacks
 
 -- | Adds a 'CallStack' to the given 'AnnotatedException'. This function will
 -- search through the existing annotations, and it will not add a second
--- 'CallStack' to the list.
+-- 'CallStack' to the list. Instead, it will append the contents of the given
+-- 'CallStack' to the existing one.
+--
+-- This mirrors the behavior of the way 'HasCallStack' actually works.
 --
 -- @since 0.1.0.0
 addCallStackToException
     :: CallStack
     -> AnnotatedException exception
     -> AnnotatedException exception
-addCallStackToException cs exn =
-    case annotatedExceptionCallStack exn of
-        Nothing ->
-            annotate [callStackToAnnotation cs] exn
-        Just _ ->
-            exn
+addCallStackToException cs (AnnotatedException annotations' e) =
+    AnnotatedException anns' e
+  where
+    anns' = go annotations'
+    -- not a huge fan of the direct recursion, but it seems easier than trying
+    -- to finagle a `foldr` or something
```

Comment:
it'd be neat if the compiler could verify that at least one branch/equation doesn't make a recursive call. that still wouldn't guarantee termination, though.

Reply:
also wouldn't catch loops in mutual recursion.